part-1.py

```diff
@@ -4,17 +4,33 @@
 # the appropriate comment.
 
 # factorial
-
-
+def factorial(n):
+    if n < 0:
+        raise ValueError
+    if n == 0 :
+        return 1
+    return n * (factorial(n - 1))
 
 # reverse
-
-
+def reverse(text):
+    if len(text) <= 1:
+        return text
+    return text[-1] + reverse(text[:-1])
 
 # bunny
-
+def bunny (count):
+    if count == 0:
+        return 0
+    return 2 + bunny(count-1)
 
 
 # is_nested_parens
+def is_nested_parens(parens):
+    if len(parens) == 0 :
+        return True
+    elif len(parens) % 2 != 0 :
+        return False
```

Comment:
This works, but it's going to short circuit the recursion. A recursive base case is something more like:
```suggestion
    elif len(parens) == 1 :
        return False
```